core/src/muxing.rs

```diff
@@ -51,178 +50,63 @@
 //! implementation of `StreamMuxer` to control everything that happens on the wire.
 
-use fnv::FnvHashMap;
-use futures::{future, prelude::*, task::Context, task::Poll};
+use futures::{task::Context, task::Poll};
 use multiaddr::Multiaddr;
-use parking_lot::Mutex;
-use std::{
-    fmt, io,
-    ops::Deref,
-    pin::Pin,
-    sync::atomic::{AtomicUsize, Ordering},
-};
+use std::collections::VecDeque;
+use std::io;
 
+pub use self::boxed::StreamMuxerBox;
+pub use self::boxed::SubstreamBox;
 pub use self::singleton::SingletonMuxer;
 
+mod boxed;
 mod singleton;
 
-/// Implemented on objects that can open and manage substreams.
-///
-/// The state of a muxer, as exposed by this API, is the following:
-///
-/// - A connection to the remote. The `poll_event`, `flush_all` and `close` methods operate
-///   on this.
-/// - A list of substreams that are open. The `poll_outbound`, `read_substream`, `write_substream`,
-///   `flush_substream`, `shutdown_substream` and `destroy_substream` methods allow controlling
-///   these entries.
-/// - A list of outbound substreams being opened. The `open_outbound`, `poll_outbound` and
-///   `destroy_outbound` methods allow controlling these entries.
+/// Provides multiplexing for a connection by allowing users to open substreams.
 ///
+/// A [`StreamMuxer`] is a stateful object and needs to be continuously polled to make progress.
 pub trait StreamMuxer {
     /// Type of the object that represents the raw substream where data can be read and written.
     type Substream;
 
-    /// Future that will be resolved when the outgoing substream is open.
-    type OutboundSubstream;
-
-    /// Error type of the muxer
-    type Error: Into<io::Error>;
-
-    /// Polls for a connection-wide event.
-    ///
-    /// This function behaves the same as a `Stream`.
-    ///
-    /// If `Pending` is returned, then the current task will be notified once the muxer
-    /// is ready to be polled, similar to the API of `Stream::poll()`.
-    /// Only the latest task that was used to call this method may be notified.
+    /// Polls the [`StreamMuxer`] to make progress.
     ///
     /// It is permissible and common to use this method to perform background
     /// work, such as processing incoming packets and polling timers.
     ///
     /// An error can be generated if the connection has been closed.
-    fn poll_event(
-        &self,
-        cx: &mut Context<'_>,
-    ) -> Poll<Result<StreamMuxerEvent<Self::Substream>, Self::Error>>;
-
-    /// Opens a new outgoing substream, and produces the equivalent to a future that will be
-    /// resolved when it becomes available.
-    ///
-    /// The API of `OutboundSubstream` is totally opaque, and the object can only be interfaced
-    /// through the methods on the `StreamMuxer` trait.
-    fn open_outbound(&self) -> Self::OutboundSubstream;
+    fn poll(&mut self, cx: &mut Context<'_>)
+        -> Poll<io::Result<StreamMuxerEvent<Self::Substream>>>;
```

Comment:
In what cases does it makes sense in poll functions to have `&mut self`, and when should it be `Pin<&mut Self>`? I wasn't sure which of the two to take in `Transport::poll`, and settled for `Pin<&mut Self>` to mirror the `Stream` trait. But it would be good to have more profound reasoning for when to use what in our API

Reply:
It is my understanding that `Pin` was introduced to allow self-referential structs in futures that are generated from async blocks. Within our hand-written state machines, we don't have self-referential structs and thus, they don't need to be pinned during execution within an executor. In other words, they will automatically implement `Unpin`.

Ideally, `Pin` wouldn't exist and all poll-functions would take `&mut self` but this wouldn't allow for safe self-referential structs within futures. Thus I think the way to go is: Use `&mut self` until the compiler complains :)

Reply:
Thank you for the explanation @thomaseizinger!

> Thus I think the way to go is: Use &mut self until the compiler complains :)

But for traits we won't really know it the struct that implements it will have self-references, no? So the compiler will not complain when defining the trait, but might complain once a user implements it on their type.

Reply:
> > Thus I think the way to go is: Use &mut self until the compiler complains :)
> 
> But for traits we won't really know it the struct that implements it will have self-references, no? So the compiler will not complain when defining the trait, but might complain once a user implements it on their type. 

That is true!

The safer way would be to use `Pin` everywhere but it is also additional complexity. I could go either way.

There are probably no implementations of `StreamMuxer` outside of this repository so leaving it at `&mut self` shouldn't cause any problems but at the same time, the additional complexity would also only really affect us core maintainers 🤷‍♂️

Reply:
Interesting! Happy to move towards `Pin<&mut self>`. I guess without knowing, I was working towards #1389 :D